lib/shrine/plugins/derivation_endpoint.rb

```diff
@@ -0,0 +1,1295 @@
+    # the derivation block:
+    #
+    #     derivation :thumbnail do |file, uploaded_file, width, height|
+    #       uploaded_file             #=> #<Shrine::UploadedFile>
+    #       uploaded_file.id          #=> "9a7d1bfdad24a76f9cfaff137fe1b5c7.jpg"
+    #       uploaded_file.storage_key #=> "store"
+    #       uploaded_file.metadata    #=> {}
+    #
+    #       # ...
+    #     end
+    #
+    # By default original metadata that was extracted during attachment isn't
+    # available, to keep the derivation URL as short as possible. However, if
+    # you want to have original metadata available when deriving, you can set
+    # the `:metadata` option to the list of needed metadata values:
+    #
+    #     plugin :derivation_endpoint, metadata: ["filename", "mime_type"]
+    #
+    # Now `filename` and `mime_type` metadata values will be available in the
+    # derivation block:
+    #
+    #     derivation :thumbnail do |file, uploaded_file, width, height|
+    #       uploaded_file.metadata #=>
+    #       # {
+    #       #  "filename" => "nature.jpg",
+    #       #  "mime_type" => "image/jpeg"
+    #       # }
+    #
+    #       uploaded_file.original_filename #=> "nature.jpg"
+    #       uploaded_file.mime_type         #=> "image/jpeg"
+    #
+    #       # ...
+    #     end
+    #
+    # ## Downloading
+    #
+    # When a derivation is requested, the original uploaded file will be
+    # downloaded to disk before the derivation block is called. If you want
+    # to pass in additional storage-specific download options, you can do so
+    # via `:download_options`:
+    #
+    #     plugin :derivation_endpoint, download_options: {
+    #       sse_customer_algorithm: "AES256",
+    #       sse_customer_key:       "secret_key",
+    #       sse_customer_key_md5:   "secret_key_md5",
+    #     }
+    #
+    # When using `Shrine.derivation_endpoint` or `Shrine.derivation_response`,
+    # if the original uploaded file has been deleted, the error the storage
+    # raises when attempting to download it will be propagated by default. You
+    # can choose to have the endpoint convert these errors to 404 responses by
+    # adding them to `:download_errors`:
+    #
+    #     plugin :derivation_endpoint, download_errors: [
+    #       Errno::ENOENT,              # raised by Shrine::Storage::FileSystem
+    #       Aws::S3::Errors::NoSuchKey, # raised by Shrine::Storage::S3
+    #     ]
+    #
+    # If you don't want the uploaded file to be downloaded to disk for you, set
+    # `:download` to `false`.
+    #
+    #     plugin :derivation_endpoint, download: false
+    #
+    # In this case the `UploadedFile` object is yielded to the derivation block
+    # instead of the raw file:
+    #
+    #     derivation :thumbnail do |uploaded_file, width, height|
+    #       uploaded_file #=> #<Shrine::UploadedFile>
+    #
+    #       # ...
+    #     end
+    #
+    # ## Derivation API
```

Comment:
A little mention on why you choose the encryption algorithm would also be nice. Any strengths/weaknesses to be aware of?

Reply:
> It would be nice to have a section on security like you did for performance.

👍 

> A little mention on why you choose the encryption algorithm would also be nice. Any strengths/weaknesses to be aware of?

It's not technically an "encryption" algorithm, as the result cannot be "decrypted" to get the original data. It's essentially a normal SHA256 hashing algorithm where secret key is used to create the hash.

I took it from Roda's `route_csrf` plugin, but Refile and `ActiveStorage::MessageVerifier` also use this. You're right that I could write a bit more about it, [like Jeremy did](https://github.com/jeremyevans/roda/blob/9c2a1fa24ffbf03ee022de62443f12ed3e3e8f38/lib/roda/plugins/route_csrf.rb#L121-L148).

Reply:
I was thinking if I were concerned about security, I'd like to know a little about the hashing, it's strength and weakness or in what case I might be concerned if you happen to know. He has a nice write up for his CSRF token.

Reply:
I added a mention for signature and DoS attack prevention. About the hashing algorithm, I feel like the documentation is already very long, and I don't think a section explaining the algorithm is useful enough. If anyone wants to find out, I'd prefer if they just look at the source code and google the algorithm.